component-monitor/src/main.rs

```diff
@@ -0,0 +1,242 @@
+                debug!(target: "files", "created {:?}", &target_path)
+            }
+        } else {
+            if let Err(err) = fs::copy(path, &target_path) {
+                error!(target: "files", "failed to copy {:?} to {:?}: {err}", &path, &target_path);
+            } else {
+                debug!(target: "files", "copied {:?} to {:?}", &path, &target_path);
+            }
+        }
+    }
+
+    if let Err(err) = fs::copy(&sentinel_src, &sentinel_tgt) {
+        error!(target: "files", "failed to copy sentinel: {err}");
+    } else {
+        debug!(target: "files", "copied sentinel ({:?} to {:?})", sentinel_src, sentinel_tgt);
+    }
+
+    Ok(())
+}
+
+/// Handle inotify events:
+/// - run cleanup on sentinel deletion
+/// - populate on sentinel written
+/// - panic on source remove
+fn monitor(
+    inotify: &Inotify,
+    source: &PathBuf,
+    sentinel_name: &str,
+    target: &PathBuf,
+) -> Result<()> {
+    info!(target: "inotify", "starting event monitoring on {:?}", &source);
+    loop {
+        let events = inotify
+            .read_events()
+            .context("Error reading Inotify events")?;
+
+        for InotifyEvent { name, mask, .. } in events {
+            debug!(
+                target: "inotify",
+                "handling {:?} event for {:?}",
+                mask,
+                name
+            );
+
+            if name.is_some_and(|n| n == sentinel_name) {
+                match mask {
+                    AddWatchFlags::IN_DELETE => cleanup(&target, &sentinel_name)?,
+                    AddWatchFlags::IN_CLOSE_WRITE => populate(&source, &sentinel_name, &target)?,
+                    _ => bail!("Unexpected Inotify event: {:?} on the sentinel file", mask),
+                }
+            } else if mask.intersects(AddWatchFlags::IN_DELETE_SELF | AddWatchFlags::IN_MOVE_SELF) {
+                bail!("Monitored folder disappeared");
+            }
+        }
+    }
+}
+
+/// Set up source directory monitoring, populate the target directory and start monitoring
+/// the source for sentinel changes
+fn main() -> Result<()> {
+    let args = Cli::parse();
+    let mut signals = Signals::new([SIGINT, SIGTERM]).context("Failed to set up signals")?;
+    let inotify = Inotify::init(InitFlags::IN_CLOEXEC).context("Failed to set up inotify")?;
+
+    let source = match args.sentinel_path.parent() {
+        Some(p) if p.is_dir() => p,
+        _ => bail!(
+            "Sentinel ({:?})'s parent is not a directory",
+            args.sentinel_path
+        ),
+    };
+
+    let sentinel_name = args
+        .sentinel_path
+        .file_name()
+        .context("Sentinel name could not be determined")?
+        .to_str()
+        .context("Sentinel name could not be read as a string")?;
+
+    if !args.target.is_dir() {
+        bail!("Target ({:?}) is not a directory", args.target);
+    }
+
+    thread::spawn(move || {
+        signals.forever().next();
+        std::process::exit(0);
+    });
```

Comment:
I guess you can also ensure here that the thread exits gracefully and logs the cause:

```suggestion
    thread::spawn(move || {
        if signals.forever().next().is_some() {
            info!("Shutting down due to signal");
            std::process::exit(0);
        }
    });
```